src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -486,46 +494,55 @@ private boolean checkSynchronized(final DetailAST synchronizedAst, boolean useBr
+        DetailAST ast;
+        if (nextCase.getType() == TokenTypes.CASE_GROUP) {
+            ast = nextCase.getFirstChild();
         }
         else {
-            final int startLineNo = currentCase.getLineNo();
-            for (int i = endLineNo - 2; i > startLineNo - 1; i--) {
-                final int[] line = getLineCodePoints(i);
-                if (!CodePointUtil.isBlank(line)) {
-                    allThroughComment = matchesComment(reliefPattern, i + 1);
-                    break;
-                }
+            ast = currentCase;
+        }
+        boolean allThroughComment = false;
+        while (ast != null) {
+            if (ast.getFirstChild() != null
+                    && reliefPattern.matcher(ast.getFirstChild().getText()).find()) {
+                allThroughComment = true;
+                break;
             }
+            ast = ast.getNextSibling();
         }
-        return allThroughComment;
+        return allThroughComment || commentAsChildOfSameGroup(reliefPattern, currentCase);
     }
 
     /**
-     * Does a regular expression match on the given line and checks that a
-     * possible match is within a comment.
+     * This method will check if the comment are child of same case group ast.
      *
-     * @param pattern The regular expression pattern to use.
-     * @param lineNo The line number in the file.
-     * @return True if a match was found inside a comment.
+     * @param reliefPattern The regular expression pattern to match the comment text
+     * @param currentCase AST of the case that falls through to the next case.
+     * @return true if relief comment found
      */
-    // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
-    @SuppressWarnings("deprecation")
-    private boolean matchesComment(Pattern pattern, int lineNo) {
-        final String line = getLine(lineNo - 1);
-
-        final Matcher matcher = pattern.matcher(line);
-        return matcher.find()
-                && getFileContents().hasIntersectionWithComment(
-                        lineNo, matcher.start(), lineNo, matcher.end());
+    private static boolean commentAsChildOfSameGroup(Pattern reliefPattern,
+                                                     DetailAST currentCase) {
+        boolean allThroughComment = false;
+        DetailAST node = currentCase.getLastChild();
+        while (true) {
+            if (node.getType() == TokenTypes.COMMENT_CONTENT
+                    && reliefPattern.matcher(node.getText()).find()) {
+                allThroughComment = true;
+            }
+            if (node.equals(currentCase.getLastChild().getLastChild())) {
+                break;
+            }
+            DetailAST visitNode = node.getFirstChild();
+            while (visitNode == null) {
+                visitNode = node.getNextSibling();
+                node = node.getParent();
+            }
+            node = visitNode;
```

Comment:
Did you mean to iterate on all comment nodes ?

Can you make function to return all by type ?
And iterate over results list of nodes for relief comment search.

Reply:
yes it is required to iterate all the node.

assume situation ex 1

```
            case 1: /* random */
                // fall through
             /* comment */
               i++;
            case 2:
                break;
```
```
|   |--CASE_GROUP -> CASE_GROUP [75:12]
    |       |   |   |--LITERAL_CASE -> case [75:12]
    |       |   |   |   |--EXPR -> EXPR [75:17]
    |       |   |   |   |   `--NUM_INT -> 1 [75:17]
    |       |   |   |   `--COLON -> : [75:18]
    |       |   |   `--SLIST -> SLIST [78:16]
    |       |   |       |--EXPR -> EXPR [78:16]
    |       |   |       |   `--POST_INC -> ++ [78:16]
    |       |   |       |       |--BLOCK_COMMENT_BEGIN -> /* [75:20]
    |       |   |       |       |   |--COMMENT_CONTENT ->  random  [75:22]
    |       |   |       |       |   `--BLOCK_COMMENT_END -> */ [75:29]
    |       |   |       |       |--SINGLE_LINE_COMMENT -> // [76:16]
    |       |   |       |       |   `--COMMENT_CONTENT ->  fall through\n [76:18]
    |       |   |       |       |--BLOCK_COMMENT_BEGIN -> /* [77:13]
    |       |   |       |       |   |--COMMENT_CONTENT ->  comment  [77:15]
    |       |   |       |       |   `--BLOCK_COMMENT_END -> */ [77:23]
    |       |   |       |       `--IDENT -> i [78:15]
    |       |   |       `--SEMI -> ; [78:18]
    |       |   |--CASE_GROUP -> CASE_GROUP [79:12]

```


case 2 ;- 
```
            case 1:
                // random
                // fall through
            // comment
                System.out.println("check");
            case 2:
```
```
|   |--CASE_GROUP -> CASE_GROUP [103:12]
    |       |   |   |--LITERAL_CASE -> case [103:12]
    |       |   |   |   |--EXPR -> EXPR [103:17]
    |       |   |   |   |   `--NUM_INT -> 1 [103:17]
    |       |   |   |   `--COLON -> : [103:18]
    |       |   |   `--SLIST -> SLIST [107:34]
    |       |   |       |--EXPR -> EXPR [107:34]
    |       |   |       |   `--METHOD_CALL -> ( [107:34]
    |       |   |       |       |--DOT -> . [107:26]
    |       |   |       |       |   |--DOT -> . [107:22]
    |       |   |       |       |   |   |--SINGLE_LINE_COMMENT -> // [104:16]
    |       |   |       |       |   |   |   `--COMMENT_CONTENT ->  random\n [104:18]
    |       |   |       |       |   |   |--SINGLE_LINE_COMMENT -> // [105:16]
    |       |   |       |       |   |   |   `--COMMENT_CONTENT ->  fall through\n [105:18]
    |       |   |       |       |   |   |--SINGLE_LINE_COMMENT -> // [106:12]
    |       |   |       |       |   |   |   `--COMMENT_CONTENT ->  comment\n [106:14]
    |       |   |       |       |   |   |--IDENT -> System [107:16]
    |       |   |       |       |   |   `--IDENT -> out [107:23]
    |       |   |       |       |   `--IDENT -> println [107:27]
    |       |   |       |       |--ELIST -> ELIST [107:35]
    |       |   |       |       |   `--EXPR -> EXPR [107:35]
    |       |   |       |       |       `--STRING_LITERAL -> "check" [107:35]
    |       |   |       |       `--RPAREN -> ) [107:42]
    |       |   |       `--SEMI -> ; [107:43]
    |       |   |--CASE_GROUP -> CASE_GROUP [108:12]

```

as you observe in both the asts without iterating all the nodes it might not be possible to detect fall thru

Reply:
Ok, please reply to second question and suggestion.

Reply:
yes i have introduce a new method as you suggested please check, 

This method will take starting node and last node as parameter and return all the `TokenTypes.COMMENT_CONTENT`
as form of list

and we can iterate in both the method hasFallThroughComment which look for comment in next case grup and in commentAsChildOfSameGroup which look for fall through in same case group to find the fallthru

I